packages/theme/src/cli/utilities/theme-environment/hmr.ts

```diff
@@ -0,0 +1,123 @@
+import {Theme} from '@shopify/cli-kit/node/themes/types'
+import {renderWarning} from '@shopify/cli-kit/node/ui'
+import EventEmitter from 'node:events'
+
+const eventEmitter = new EventEmitter()
+const updatedReplaceTemplates = {} as {[key: string]: string}
+
+type HmrEvent =
+  | {
+      type: 'section'
+      key: string
+      content: string
+    }
+  | {
+      type: 'other'
+      key: string
+    }
+
+function emitHmrEvent(event: HmrEvent) {
+  eventEmitter.emit('hmr', event)
+}
+
+export function getReplaceTemplates() {
+  return {...updatedReplaceTemplates}
+}
+
+export function setReplaceTemplate(key: string, content: string) {
+  updatedReplaceTemplates[key] = content
+}
+
+export function deleteReplaceTemplate(key: string) {
+  delete updatedReplaceTemplates[key]
+}
+
+export function getHmrHandler() {
+  return defineEventHandler((event) => {
+    if (event.path !== '/__hmr') return
+
+    const eventStream = createEventStream(event)
+
+    eventEmitter.on('hmr', (event: HmrEvent) => {
+      eventStream.push(JSON.stringify(event)).catch((error: Error) => {
+        renderWarning({headline: 'Failed to send HMR event.', body: error?.stack})
+      })
+    })
+
+    return eventStream.send()
+  })
+}
+
+export async function triggerHmr(theme: Theme, ctx: DevServerContext, key: string) {
+  const sectionId = key.match(/^sections\/(.+)\.liquid$/)?.[1]
+
+  if (!sectionId) {
+    emitHmrEvent({type: 'other', key})
+    return
+  }
+
+  const sectionTemplate = updatedReplaceTemplates[key]
+  if (!sectionTemplate) {
+    renderWarning({headline: 'No template found for HMR event.', body: `Template ${key} not found.`})
+    return
+  }
+
+  const response = await render(ctx.session, {
+    path: '/',
+    query: [],
+    themeId: String(theme.id),
+    cookies: '',
+    sectionId,
+    headers: {},
+    replaceTemplates: {[key]: sectionTemplate},
+  })
+
+  const content = await response.text()
+
+  emitHmrEvent({type: 'section', key, content})
+}
+
+function injectFunction(fn: () => void) {
+  return `<script>(${fn.toString()})()</script>`
+}
+
+export function injectFastRefreshScript(html: string) {
+  // These function runs in the browser:
+
+  function fastRefreshScript() {
+    const evtSource = new EventSource('/__hmr', {withCredentials: true})
+
+    evtSource.onopen = () => {
+      // eslint-disable-next-line no-console
+      console.info('[HMR] Connected')
+    }
+
+    evtSource.onmessage = (event) => {
+      if (typeof event.data !== 'string') return
+
+      const data = JSON.parse(event.data) as HmrEvent
+      if (data.type === 'section') {
+        const id = data.content.match(/id="([^"]+)"/)?.[1]
+        if (id) {
+          const existingElement = document.getElementById(id)
+          if (existingElement) {
+            existingElement.outerHTML = data.content
+            // eslint-disable-next-line no-console
+            console.info('[HMR] Updated section:', data.key)
```

Comment:
The same section file may impact multiple DOM elements, as we may notice here:

![Screenshot 2024-08-12 at 09 49 10](https://github.com/user-attachments/assets/530580a2-43f1-46f4-b724-08b500ad1f79)

The `__SHOPIFY_CLI_ENV__` global variable, that supports the process above, is injected/created here:
https://github.com/Shopify/cli/blob/142797ebc1a260936762d4cddcc88102d6c72404/packages/cli-kit/assets/cli-ruby/lib/shopify_cli/theme/dev_server/hot_reload/script_injector.rb#L39-L48

(that variable helps the client side on the mapping of the expected DOM elements)